### PR TITLE
[FLINK-4443][rpc] Add support in RpcCompletenessTest for inheritance of RpcGateway and RpcEndpoint

### PR DESCRIPTION
**Pull request by wenlong88**
RpcCompletenessTest needs to support RpcGateway which is composited by some basic functions like following:
```
public interface ExecutionStateListener extends RpcGateway {
	public void notifyExecutionStateChanges();
}
public interface JobStateListener extends RpcGateway {
	public void notifyJobStateChanges();
}
public interface JobWatcher extends ExecutionStateListener, JobStateListener, RpcGateway {

}
public class JobWatcherEndpoint extends RpcEndpoint<JobWatcher> {
	protected JobWatcherEndpoint(RpcService rpcService) {
		super(rpcService);
	}
	@RpcMethod
	public void notifyExecutionStateChanges() {

	}
	@RpcMethod
	public void notifyJobStateChanges() {

	}
}
public class AttachedJobClient extends JobWatcherEndpoint {
	protected JobClient(RpcService rpcService) {
		super(rpcService);
	}
}
```

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/RpcGateway.java [hunk 1]
```diff
@@ -28,5 +28,6 @@
 	 *
 	 * @return Fully qualified address under which the associated rpc endpoint is reachable
 	 */
+	@NativeMethod
 	String getAddress();
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/RpcGateway.java]: Can't we simply say that all methods defined in `RpcGateway` are implicitly a *native* method. Then we could save the `@NativeMethod` annotation.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcCompletenessTest.java [hunk 2]
```diff
@@ -337,4 +337,25 @@ private static boolean isRpcTimeout(Annotation[] annotations) {
 			throw new RuntimeException("Could not retrive basic type information for primitive type " + primitveType + '.');
 		}
 	}
+
+	/**
+	 * Extract all rpc methods defined by the gateway interface
+	 * @param interfaceClass the given rpc gateway interface
+	 * @return all methods defined by the given interface
+	 */
+	private List<Method> getRpcMethodsFromGateway(Class<? extends RpcGateway> interfaceClass) {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcCompletenessTest.java]: Here we would simply test if `interfaceClass` is of type `RpcGateway` and if so, then return an empty list.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcCompletenessTest.java [hunk 3]
```diff
@@ -337,4 +337,25 @@ private static boolean isRpcTimeout(Annotation[] annotations) {
...
+		if(!interfaceClass.isInterface()) {
+			fail(interfaceClass.getName() + "is not a interface");
+		}
+		ArrayList<Method> allMethods = new ArrayList<>();
+		for(Method method : interfaceClass.getDeclaredMethods()) {
+			if (!method.isAnnotationPresent(NativeMethod.class)) {
+				allMethods.add(method);
+			}
+		}
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcCompletenessTest.java]: line break missing

**Comment by tillrohrmann** [comment]
Thanks for your contribution @wenlong88. Your changes look good to me. I only had a comment concerning the `NativeMethod` annotation. I think that at the moment we could get along without it by considering all `RpcGateway` methods implicitly native. That would simplify decrease the complexity a little bit because we can get rid of/not introduce another annotation. What do you think?

**Comment by wenlong88** [comment]
@tillrohrmann  all comments addressed. I have removed the `NativeMethod` annotation.

**Comment by tillrohrmann** [comment]
Great @wenlong88. Changes look good to me. Will merge your PR. Thanks a lot for your contribution :-) Please close the PR after I've merged it.

**Pull request closed without merging**